// libmuse/script/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.use_proxy = exports.sum_total_duration = exports.j = exports.jo = exports.debug = exports.wait = void 0;
const deps_js_1 = require("./deps.js");
const errors_js_1 = require("./errors.js");
const setup_js_1 = require("./setup.js");
/**
 * Wait a given number of milliseconds, then resolve
 */
const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
exports.wait = wait;
/**
 * Logs debug messages to the console
 */
const debug = (...args) => {
    if ((0, setup_js_1.get_option)("debug"))
        console.debug(...args);
};
exports.debug = debug;
const jo = (json, path, ...others) => {
    const result = (0, deps_js_1.JSONPath)({ path: [path, ...others].join("."), json });
    return result.length ? result[0] : null;
};
exports.jo = jo;
const j = (json, path, ...others) => {
    const result = (0, exports.jo)(json, path, ...others);
    if (!result) {
        throw new errors_js_1.MuseError(errors_js_1.ERROR_CODE.PARSING_INVALID_JSON, `JSONPath expression "${[path, ...others]}" returned nothing`);
    }
    return result;
};
exports.j = j;
function sum_total_duration(item) {
    if (!("tracks" in item)) {
        return 0;
    }
    else {
        return item.tracks.reduce((acc, track) => acc + track.duration_seconds || 0, 0);
    }
}
exports.sum_total_duration = sum_total_duration;
function use_proxy(url) {
    const proxy = (0, setup_js_1.get_option)("proxy");
    if (proxy) {
        return proxy + "/" + new URL(url).href;
    }
    else {
        return url;
    }
}
exports.use_proxy = use_proxy;
